scripts/build_utils.sh

```diff
@@ -1694,54 +1690,157 @@ pr_only_for() {
   fi
   if [ -f $(git rev-parse --git-dir)/shallow ]; then
     # We can't do a regular `git diff` in a shallow clone.  There is no other way to check files changed.
-    files="$(curl -s -u ${GITHUB_USER}:${GITHUB_PASS} https://api.github.com/repos/${ghprbGhRepository}/pulls/${ghprbPullId}/files | jq '.[].filename' | tr -d '"')"
+    curl -s -u ${GITHUB_USER}:${GITHUB_PASS} https://api.github.com/repos/${ghprbGhRepository}/pulls/${ghprbPullId}/files | jq '.[].filename' | tr -d '"'
   else
-    files="$(git diff --name-only origin/${ghprbTargetBranch}...origin/pr/${ghprbPullId}/head)"
+    git diff --name-only origin/${ghprbTargetBranch}...origin/pr/${ghprbPullId}/head
   fi
   popd
-  echo -e "changed files:\n$files"
-  # 0 is true, 1 is false
-  local all_match=0
+}
+
+# construct a regex pattern that matches the CODEOWNERS syntax rules
+# for ? * and **. these rules use a subset of the gitignore pattern format
+# https://git-scm.com/docs/gitignore#_pattern_format
+codeowners_pattern_to_regex() {
+  # first escape any regex characters like \^()[]|.+ before we add more
+  local escape='s/[\\\^\(\)\[\]\|\.\+]/\\&/g'
+  # replace ** patterns with ++ until we replace single *s
+  # /**/ -> /++
+  local slash_double_asterisk_slash='s/\/\*\*\//\/++/g'
+  # /** -> /++
+  local slash_double_asterisk='s/\/\*\*/\/++/g'
+  # **/ -> ++/
+  local double_asterisk_slash='s/\*\*\//++\//g'
+  # * matches any characters except slashes: [^/]*
+  local asterisk='s/\*/[^\/]*/g'
+  # ? matches any character except slash: [^/]
+  local question='s/\?/[^\/]/g'
+  # ++ (aka **) matches anything: .*
+  local double_plus='s/\+\+/.*/g'
+  echo $1 | sed \
+      -e $escape \
+      -e $slash_double_asterisk_slash \
+      -e $double_asterisk_slash \
+      -e $slash_double_asterisk \
+      -e $asterisk \
+      -e $question \
+      -e $double_plus
+}
+
+codeowners_patterns_to_regex() {
+  for p in $1; do
+    # skip comments
+    if [[ $p == \#* ]]; then continue; fi
+    codeowners_pattern_to_regex $p
```

Comment:
from https://jenkins.ceph.com/view/all/job/preserve-ceph-pull-requests/2/consoleFull for https://github.com/ceph/ceph/pull/63078

given input like:
```
++ codeowners_patterns_to_regex 'doc/
admin/
src/sample.ceph.conf
CodingStyle
*.rst
*.md
COPYING*
README.*
SubmittingPatches
.readthedocs.yml
PendingReleaseNotes'
```
i see patterns like `README.*` being expanded where they shouldn't
```
++ for p in $1
++ [[ README.windows.rst == \#* ]]
++ codeowners_pattern_to_regex README.windows.rst
```